Initial support for RPM packages

brakarov opened the pull request:
This should be a start towards fixing #5, since dnf and yum should be compatible. Will probably add explicit yum support later.

I would like to implement a dnf_repo which handles new yum repos. Since these are ini-files I would suggest taking the relevant code out of file_ini, refactor it and move into functions/ini_file. Are you okay with this?

By the way, feedback is welcome. This is a new codebase and I've tried following your coding conventions but you're the expert.

jtopjian commented:
I did a quick review and this looks awesome! Very nice work.

> I would like to implement a dnf_repo which handles new yum repos. Since these are ini-files I would suggest taking the relevant code out of file_ini, refactor it and move into functions/ini_file. Are you okay with this?

I'd say go ahead and explore this and see how it looks in the end. Another option would be to make `dnf_repo` sort of a wrapper around `file_ini` like what is done in the [mysql.mycnf](https://github.com/wffls/waffles/blob/master/functions/mysql.sh) function.

> By the way, feedback is welcome. This is a new codebase and I've tried following your coding conventions but you're the expert.

Far from it. :smile: 

But again, this PR looks really good. I recently ripped a lot of stuff out of Waffles and need to document how the internals work. It looks like you caught on really quick, which I'm happy to see!

brakarov commented:
> Another option would be to make dnf_repo sort of a wrapper around file_ini like what is done in the mysql.mycnf function.

I've thought about that as well. But wouldn't any errors be reported as file_ini instead of dnf_repo? If this happens in a script with another file_ini this would confuse the user.

jtopjian commented:
That's a good point. I'd like to be able to use some of the more basic resources as building blocks for other resources, though I haven't thought that all the way through.

The `$waffles_resource` variable is set [here](https://github.com/wffls/waffles/blob/master/resources/file_ini.sh#L25) in each resource. It's then used for logging [here](https://github.com/wffls/waffles/blob/master/functions/logging.sh#L30).

I wonder if either resources should be "stacked" or only log the outermost `$waffles_resource` when resources are nested.

brakarov commented:
I wouldn't stack resources. This takes away control from the outermost resource. I'll give an example: `res.example --opt attr` is an optional attribute and if file_ini for whatever reason tries to change this and it fails it would throw an error. The developer of res.example would prefer a warning in this case. I don't think this could be achieved with stacking resources (unless you start using > /dev/null tricks).

It's better, I think, to abstract the underlying functionality. file_ini is now neatly split into two parts:

1. **What** needs to be done: Everything above [ini_get_option](https://github.com/wffls/waffles/blob/master/resources/file_ini.sh#L98)
2. **How** it is done: Everything below. This would end up in functions/

I'll try to push a commit later today with file_ini split up and moved into functions so you can see how it looks.

jtopjian commented:
Is it possible to rebase this branch against an updated master?

brakarov commented:
Apperently not. I've done something wrong I guess, this is the first time I am collaborating with git. I'll close and fork again. Unless you know how to clean this mess up?

jtopjian commented:
Yup, it's pretty simple. 

If you want to have some safety, either:
* Just make a copy of the entire waffles directory (`cp ~/waffles ~/waffles.orig`). That will copy the git state as well, so if something really blows up, just trash the entire directory and try again.
* Check out a new branch off of fedora (`git checkout -b fedora-new`). Then you should be able to return to the original fedora branch.

To rebase:
1. First, switch to the master branch and do a `git pull` to bring in the latest commits.
2. Next, switch to your fedora branch. 
3. Run `git rebase -i master`

That _should_ figure out what commits are unique to the fedora branch and overlay them on master. Things get a little messy if you run into a merge conflict.

Once everything is rebased, you can overwrite the upstream fedora branch by doing `git push -f -u brakarov fedora`. This PR will automatically have the changes reflected.

I've only been working with git in a collaborative way for a year and ran into this kinds of problems, too. Play around, make a backup, break things, and let me know if you need a hand. :)

brakarov commented:
Required the use of several git push -f but this looks a lot better.

jtopjian commented on resources/dnf_copr.sh:
```diff
@@ -0,0 +1,78 @@
+  fi
+
+  # Check if all dependencies are installed
+  local _wrd=("${_dnf_cmd}" "grep" "cut" "sed")
```
Since coreutils are an implicit requirement, `cut` can be excluded here.

jtopjian commented on resources/dnf_pkg.sh:
```diff
@@ -0,0 +1,149 @@
+  fi
+
+  # Check if all dependencies are installed
+  local _wrd=("${_dnf_cmd}" "rev" "cut" "tr" "tail")
```
Since coreutils is an implicit requirement, `cut`, `tr`, and `tail` can be excluded here. On Ubuntu, `rev` is part of a `util-linux` package... let's keep `rev` here for now.

jtopjian commented:
Nice 😄 

Two quick comments are requirements.

I also have a question about `dnf.pkg`: is it logical for someone to specify both an `*.rpm` file _and_ a version? Should an error be presented if those two combinations are used?

brakarov commented:
> I also have a question about dnf.pkg: is it logical for someone to specify both an *.rpm file and a version? Should an error be presented if those two combinations are used?

You're right. I've updated this. Is it the correct place in read(), because this could be checked sooner in the main function.

Both comments are updated. Thank you for checking what's in coreutils!

I've also added the dnf_repo resource. This should complete what I think is needed for proper dnf/yum supoprt.

Thank you for reviewing this!

jtopjian commented on resources/dnf_copr.sh:
```diff
@@ -0,0 +1,78 @@
+  local _repo_file_name="_copr_$(echo $_copr | sed -e "s|[/:]|-|" -e "s|\.|_|").repo"
+  if [ -f /etc/yum.repos.d/$_repo_file_name ]; then
+    local _is_enabled=$(grep ^enabled= /etc/yum.repos.d/$_repo_file_name | cut -d= -f2)
+    if [[ _is_enabled -eq 1 ]]; then
```
`$_is_enabled`?

jtopjian commented:
I did a code review and spotted one little error. Other than that, this looks really good :smile: 

I can work on getting a fedora container into the kitchen suite once this is merged. 

I noticed `dnf.repo` has some `noop` stuff and now just noticed that `file.ini` has it as well. I wonder if this can be moved into the new ini-functions? This doesn't have to be done now - I'm OK with how things are currently.

If you wanted some more git practice, you could try squashing these 9 commits into two: one for the new ini functions and the other for all of the dnf resources. This is done very similar to the other rebasing you did:

* `git rebase -i HEAD~9`
* reorder the commits so that all ini functions are together and all dnf resources are together
* either finish rebasing and then do another `git rebase -i HEAD~9` or just
* Change `pick` to `s` (for squash) for all commits except for one ini function and one dnf. Basically you're picking one commit to roll the other commits into.
* Finally, do one last rebase `git rebase -i HEAD~2` and edit the commit message to give a more broad log of what the two commits represent.

I found that when I started with git, all documentation made one feel like they should never use rebase, however, it's used extensively for contributing to github-based projects.

If you don't want to go through the hassle, no big deal - I'm happy to commit all of these, but I thought this might be a good test run if you run into another project where clean commits are required.

brakarov commented:
> If you wanted some more git practice, you could try squashing these 9 commits into two: one for the new ini functions and the other for all of the dnf resources.

Done. I start to understand why everybody loves git so much.  :smile:

> I did a code review and spotted one little error. Other than that, this looks really good :smile:

Thanks! Nice spotting by the way, it's fixed in the latest commit, which I've then squashed.

> I noticed dnf.repo has some noop stuff and now just noticed that file.ini has it as well. I wonder if this can be moved into the new ini-functions? This doesn't have to be done now - I'm OK with how things are currently.

Agreed. I'll open a new issue for this.

> I can work on getting a fedora container into the kitchen suite once this is merged.

 :+1: I wanted to add these as well, but lost my motivation when I saw all the dependencies I needed to install. So I opened #181...

jtopjian commented:
> Is this enough to close #5?

Absolutely! Very nice work! Thank you for taking the time to do this.
